File: CCC/CCC2009S4.py
import sys;
sys.stdin = open('s4.4.in')
input = sys.stdin.readline
n = int(input())
t = int(input())

adj = [[-1 for i in range(n+1)]for i in range(n+1)]

for i in range(t):
    a,b,d = map(int,input().split())
    if adj[a][b] == -1:
        adj[a][b] = d
        adj[b][a] = d
    else:
        adj[a][b] = min(adj[a][b],d)
        adj[b][a] = min(adj[b][a],d)

k = int(input())
pcit = []
for i in range(k):
    pcit.append(list(map(int,input().split())))

home = int(input())    

dist = [float('inf') for i in range(n+1)]
v = [False for i in range(n+1)]
dist[home] = 0
print("DONE READING")
mindex = -1
#dijkstras algorithm
for i in range(n):
    mindex = -1
    for j in range(1,n+1):
        if (not v[j]) and (mindex == -1 or (dist[mindex]>dist[j])):
            mindex = j
    #print(mindex,dist[mindex])     
    v[mindex] = True
    for edg in range(1,n+1):
        if not adj[mindex][edg] == -1:
            if dist[edg] > dist[mindex] + adj[mindex][edg]:
                dist[edg] = dist[mindex] + adj[mindex][edg]
     

best = float('inf')        
for cit in pcit:
    c,cost = cit
    best = min(best,dist[c]+cost)
    
print(best)


    

File: CCC/CCC2009S4.cpp
/*
 * CCC2009S4.cpp
 *
 *  Created on: May 21, 2017
 *      Author: Rob
 */

#include<iostream>
#include<list>
#include<algorithm>
#define pb push_back
const int INF = 1e9;
using namespace std;


int con[5001][5001];
int dist[5001];
bool v[5001] = {0};

int main(){
	int n,t,k,d;
	scanf("%d%d",&n,&t);
	int a,b,w;
	for(int i = 0; i<=5000; i++){
		for(int ii = 0; ii<=5000; ii++){
			con[i][ii] = -1;
		}
	}
	for(int i = 0; i<t ; i++){
		scanf("%d%d%d",&a,&b,&w);
		if(con[a][b] == -1){
			con[a][b] = w;
			con[b][a] = w;
		}
		else{
			con[a][b] = min(con[a][b],w);
			con[b][a] = min(con[b][a],w);
		}
	}
	for(int i = 0; i<=5000; i++){
		dist[i] = INF;
	}
	scanf("%d",&k);
	for(int i = 0; i<k; i++){
		scanf("%d%d",&a,&w);
		dist[a] = w;
	}
	scanf("%d",&d);

	//Dijkstra's Algorithm
	for(int i = 0; i<n; i++){
		int mindex = -1;
		for(int j = 0; j<=n; j++){
			if(!v[j] && (mindex == -1 || dist[j] < dist[mindex])){
				mindex = j;
			}
		}

		v[mindex] = true;
		if(mindex == d){break;}
		for(int i = 1; i<=5000; i++){
			if(con[mindex][i] == -1){continue;}
			else{dist[i] = min(dist[mindex]+con[mindex][i],dist[i]);}
		}
	}
	printf("%d",dist[d]);
	return 0;
}
